File: backend/routers/template.py
import io
import json
import asyncio
import time
from fastapi import APIRouter, UploadFile, File, Form, HTTPException, Depends
from fastapi.responses import StreamingResponse
from backend.core.config import logger, executor
from backend.core.auth import get_current_user, log_activity
from tools.template_mapper import get_excel_headers, map_template_data, preview_mapped_data

router = APIRouter(prefix="/api/file", tags=["template"])

@router.post("/template-headers")
async def get_template_headers_api(file: UploadFile = File(...)):
    try:
        is_csv = file.filename.lower().endswith(".csv")
        headers = get_excel_headers(file.file, is_csv=is_csv)
        return {"headers": headers}
    except Exception as e:
        logger.error(f"Error getting headers: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/template-preview")
async def preview_mapping_api(
    template_headers: str = Form(...),
    data_file: UploadFile = File(...),
    mapping_json: str = Form(...) 
):
    try:
        t_headers = json.loads(template_headers)
        mapping = json.loads(mapping_json)
        contents = await data_file.read()
        buffer = io.BytesIO(contents)
        is_csv = data_file.filename.lower().endswith(".csv")
        preview = preview_mapped_data(t_headers, buffer, is_csv, mapping)
        return preview
    except Exception as e:
        logger.error(f"Error in preview: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/template-map")
async def map_template_api(
    template_headers: str = Form(...),
    data_file: UploadFile = File(...),
    mapping_json: str = Form(...),
    user=Depends(get_current_user)
):
    try:
        t_headers = json.loads(template_headers)
        mapping = json.loads(mapping_json)
        contents = await data_file.read()
        buffer = io.BytesIO(contents)
        is_csv = data_file.filename.lower().endswith(".csv")
        loop = asyncio.get_running_loop()
        output, filename = await loop.run_in_executor(
            executor,
            lambda: map_template_data(t_headers, buffer, is_csv, mapping)
        )
        if output is None:
            raise HTTPException(status_code=400, detail=filename)
        if user:
            await log_activity(user.id, "Template Mapping", filename)
        return StreamingResponse(
            output,
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={"Content-Disposition": f'attachment; filename="mapped_output_{int(time.time())}.xlsx"'}
        )
    except Exception as e:
        logger.error(f"Error in template mapping: {e}")
        raise HTTPException(status_code=500, detail=str(e))

File: server-node/src/routers/template.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const upload = multer({ storage: multer.memoryStorage() });
const { getExcelHeaders, processTemplateMap, previewTemplateMap } = require('../utils/tools');
const { logActivity } = require('../core/auth');
const { getFilesFromRequest } = require('../utils/helpers');

router.post('/template-headers', upload.single('file'), async (req, res) => {
    try {
        const flatFiles = await getFilesFromRequest(req);
        if (flatFiles.length === 0) return res.status(400).json({ error: "No file provided" });
        const file = flatFiles[0];
        const headers = getExcelHeaders(file.buffer, file.originalname.toLowerCase().endsWith('.csv'));
        res.json({ headers });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

router.post('/template-preview', upload.single('data_file'), async (req, res) => {
    try {
        const flatFiles = await getFilesFromRequest(req);
        if (flatFiles.length === 0) return res.status(400).json({ error: "No data file provided" });
        const file = flatFiles[0];

        const { template_headers, mapping_json } = req.body;
        const tHeaders = JSON.parse(template_headers);
        const mapping = JSON.parse(mapping_json);

        const result = previewTemplateMap(file.buffer, {
            template_headers: tHeaders,
            mapping,
            isCsv: file.originalname.toLowerCase().endsWith('.csv')
        });

        res.json(result);
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

router.post('/template-map', upload.single('data_file'), async (req, res) => {
    try {
        const flatFiles = await getFilesFromRequest(req);
        if (flatFiles.length === 0) return res.status(400).json({ error: "No data file provided" });
        const file = flatFiles[0];

        const { template_headers, mapping_json } = req.body;
        const tHeaders = JSON.parse(template_headers);
        const mapping = JSON.parse(mapping_json);

        const { buffer, extension } = await processTemplateMap(file.buffer, {
            template_headers: tHeaders,
            mapping,
            isCsv: file.originalname.toLowerCase().endsWith('.csv')
        });

        if (req.user) {
            await logActivity(req.user.id, "Template Mapping", `mapped_output${extension}`);
        }

        res.setHeader('Content-Disposition', `attachment; filename="mapped_output_${Date.now()}${extension}"`);
        res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
        res.send(buffer);
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

module.exports = router;
